src/slurmdbd_ops.py

```diff
@@ -0,0 +1,448 @@
+
+        Note: variable names will be converted to upper case when being written
+        to the file and for comparison of keys.
+
+        Args:
+            **kwargs:
+                The environment variables that should be set or unset. The key
+                will be upper-cased for the environment variable.
+        """
+        with open(SLURMDBD_DEFAULTS, "w+") as f:
+            updated_contents = []
+
+            keys_processed = set()
+
+            for line in f.readlines():
+                # Lines beginning with a hash are a comment. Don't process these
+                # but do add the line to the output buffer to preserve it.
+                if line.startswith("#"):
+                    updated_contents.append(line)
+                    continue
+
+                # Attempt to get the environment variable and split it on the =.
+                # If the line doesn't have an equals, then don't process the line
+                # but do add the line to the output buffer to preserve it.
+                line_parts = line.split("=", 1)
+                if len(line_parts) < 2:
+                    updated_contents.append(line)
+                    continue
+
+                var_name = line_parts[0].lower()
+                if var_name not in kwargs:
+                    updated_contents.append(line)
+                    continue
+
+                keys_processed.add(var_name)
+                # If explicitly None, then remove from the output buffer as
+                # the demand is to unset the variable.
+                if kwargs[var_name] is None:
+                    continue
+
+                updated_contents.append(f"{var_name.upper()}={kwargs[var_name]}\n")
+
+            for key, value in kwargs.items():
+                # Skip any keys already processed.
+                if key in keys_processed:
+                    continue
+                updated_contents.append(f"{key.upper()}={value}\n")
+
+            f.seek(0)
+            f.truncate()
+            f.writelines(updated_contents)
+
+    @property
+    def hostname(self) -> str:
+        """Return the hostname."""
+        return socket.gethostname().split(".")[0]
+
+    @property
+    def _slurm_systemd_service(self) -> str:
+        """Return the Slurm systemd unit file."""
+        return "slurmdbd.service"
+
+    @property
+    def _munged_systemd_service(self) -> str:
+        """Return the name of the Munge Systemd unit file."""
+        return "munge.service"
+
+    @property
+    def version(self) -> str:
+        """Return slurmdbd version."""
+        return Slurmdbd().version()
+
+
+CHARM_MAINTAINED_PARAMETERS = {
+    "DbdPort": "6819",
+    "AuthType": "auth/munge",
+    "AuthInfo": '"socket=/var/run/munge/munge.socket.2"',
+    "SlurmUser": "slurm",
+    "PluginDir": "/usr/lib/x86_64-linux-gnu/slurm-wlm",
+    "PidFile": "/var/run/slurmdbd.pid",
+    "LogFile": "/var/log/slurm/slurmdbd.log",
+    "StorageType": "accounting_storage/mysql",
+}
```

Comment:
Should this possibly be in a `constants.py` module? This is a common design pattern that I have seen within other charms.